### ISSUE TITLE
Separating data collecting and management

### ISSUE THREAD

**Issue by Avamander** [Freeyourgadget/Gadgetbridge#963]
Wouldn't it be a good idea to separate Fitness/Health (Sleep, etc.) logic and device communication logic? Basically turning Gadgetbridge more bridge-ish so that it just syncs with devices, allows management of devices and collects data. Then creating something like Gadgethealth that takes the data and displays it.

This would provide for example:
* Easier to support for new devices given there's a standard API that has to be provided, wouldn't also mean a large codebase growth with every added device
* Wouldn't force anyone to be stuck with one data management/viewing interface
    * Easier to add other endpoints - OwnCloud, app, saving to disk, Syncthing and etc.
    * If there's no need for those features it wouldn't force one to spend disk space, RAM etc. on them - for example when one doesn't need OwnCloud support
    * Would allow 3rd party apps to request data from Gadgetbridge
* Reduced code complexity, smaller codebase to manage per app

Downsides:
* APIs has to stay in sync for compatibility
* Additional documentation has to be written
* API security has to be figured out, but considering that both Pebble external "timeline", background JS (with granular privacy controls) and future external data source support should be added this kind of functionality should be implemented anyways

What do you think, does this idea have any merit?

**Comment by cpfeiffer** [comment]
Some counter questions and remarks:
1. would you be willing to work on this?
1. most of our code is not very complex. Separating anything out will not make it less complex.
1. integration with third party apps is something I'd really love to see (I think the Nextcloud guys are still cooking something?)
1. we already have APIs and infrastructure that all the different devices use. Most of the UI code is device independent, there's nothing to gain from separation.

**Comment by Avamander** [comment]
> would you be willing to work on this?

I guess.

> most of our code is not very complex. Separating anything out will not make it less complex.

It's not complex, but it could be made less complex I think

> we already have APIs and infrastructure that all the different devices use. Most of the UI code is device independent, there's nothing to gain from separation.

Not accessible from 3rd party apps nicely though. Neither is there any Pebble Timeline.

**Comment by NotAFile** [comment]
I dislike the idea of having a seperate app just for the graphs. I do however think some kind of external API that allows you to interact with the gadgetbridge devices and sensors would be very neat. Particularly, some kind of way of doing integration with phone automation apps.

**Comment by Avamander** [comment]
@NotAFile Not exactly just for the graphs. I think data analysis could be finally done then too, GB currently collects (and could collect, battery prediction for ex. !) a lot but does so little with it. If that functionality is separated from GB it would also mean there's a mature API for HASS and similar solutions.

**Comment by NotAFile** [comment]
Yes, I think retaining the current functionality and allowing API access is a better solution though. Splitting the apps just makes many things more complicated. There's now two settings, two apps, two everything. You can't just press "sync" and click on the notification to see how much you slept anymore.

**Comment by Avamander** [comment]
@NotAFile 
> You can't just press "sync" and click on the notification to see how much you slept anymore.

Why not?

**Comment by Avamander** [comment]
Though I absolutely get that it would be _quite_ a bit of work, but I feel the UNIX philosophy that we should "Write programs that do one thing and do it well. Write programs to work together. [...]". 

Though maybe start maybe statistics could be turned universal, not per-device (but with an optional device filter)? Though if there's a wish to expose the data to other apps some confirmation and granular API system needs to be built anyways that also means the graphs should use that API which pretty much means it could already be decoupled.

**Comment by cpfeiffer** [comment]
@Avamander 

> It's not too complex, but it could be made less complex I think

If you have some particular code in mind, do tell.

My point is that the device specific code is already pretty much decoupled from everything else. E.g. @jpbarraca managed to implement full support for multiple devices without changing any "application" code.

> Not accessible from 3rd party apps nicely though. 

We can add that without tearing the entire app apart.

> Neither is there any Pebble Timeline.

That wouldn't magically appear when we'd split Gadgetbridge into two apps. Someone just has to do it, ideally in a device independent way.

> Though maybe at start statistics could be turned universal, not per-device (but with an optional device filter)?

Yes, that would be nice. A general, device independent statistics dashboard would be a great start!

**Comment by Avamander** [comment]
@cpfeiffer 
>> Not accessible from 3rd party apps nicely though.

> We can add that without tearing the entire app apart.

I've spent a bit of time reading the open/closed issues and I have a feeling that a good API is really the only way of actually providing good sleep, weight, activity and etc. data management people have requested, I think each of those tasks alone deserve a separate application. Good proof of that is simply how many "* trackers" there are on Google Play store. 

I do now think that totally stripping out the graphing part wouldn't be that great of an idea, advanced features aren't always what people want and just seeing step count (etc.) is sometimes enough.

What do you think? What could be done to create this sort of API other apps would want to use but still kept data as private as the users want?

**Comment by danielegobbetti** [comment]
In the last days nextcloud released the "talk" application to fdroid, and it comes with an option to import your user credentials from the main app. I didn't look their code yet, but I am fairly sure the data exchange is protected/validated somehow. I believe we can learn from their approach. If someone could look into how the main nextcloud app integrates with the "talk" app, that would be great!

**Comment by NotAFile** [comment]
@danielegobbetti I just tried this and it just opened a webview that allowed me to login to my server and grant oauth

**Comment by NotAFile** [comment]
@danielegobbetti I can't do that and I have both apps installed :shrimp:

**Comment by danielegobbetti** [comment]
@NotAFile the functionality is definitely there ( https://github.com/nextcloud/talk-android/blob/master/app/src/main/java/com/nextcloud/talk/utils/AccountUtils.java, https://github.com/nextcloud/talk-android/blob/master/app/src/main/java/com/nextcloud/talk/controllers/AccountVerificationController.java ). 

As I stated above, I would like to dig deeper to understand how they are doing it, as I trust nextcloud for doing it in a secure way.

**Comment by ildar** [comment]
> I dislike the idea of having a seperate app just for the graphs

This is correct. But IMHO _expanding_ GB to such functionality isn't a great idea.
I was thinking of FRs for better graphs and I couldn't convince myself to file my ideas into this issue tracker. GB isn't a proper app for this.
Anyone knows a kinda "Health tracking" good FOSS app? we could try to pull GB data and have all kinds of fancy graphs there.
